src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,361 @@
+                    Some(common_ip)
+                } else {
+                    None
+                }
+            )
+    }
+
+    // Get ports list of given IP
+    fn get_nat_ports(addrs: &[SocketAddr], ip: IpAddr) -> Vec<u16> {
+        addrs.iter()
+            .filter(|addr| addr.ip() == ip)
+            .map(|addr| addr.port())
+            .collect::<Vec<u16>>()
+    }
+
+    // punch using first_punching_index
+    fn first_hole_punching(&mut self, ports: Vec<u16>, ip: IpAddr, server: &Server, friend_pk: PublicKey) -> IoFuture<()> {
+        let num_ports = ports.len();
+        let first_punching_index = self.first_punching_index;
+        let ping_sender = (0..MAX_PORTS_TO_PUNCH)
+            .map(|i| {
+                let it = i + first_punching_index;
+                let sign: i16 = if it % 2 == 1 { -1 } else { 1 };
+                let delta = sign * (it / (2 * num_ports as u32)) as i16;
+                let index = (it as usize / 2) % num_ports;
+                let port = (ports[index] as i16 + delta) as u16;
+
+                server.send_ping_req(
+                    &PackedNode::new(false, SocketAddr::new(ip, port), &friend_pk)
+                )
+            });
+
+        let ping_stream = stream::futures_unordered(ping_sender).then(|_| Ok(()));
+
+        Box::new(ping_stream.for_each(|()| Ok(())))
+    }
+
+    // do punch using last_punching_index
+    fn last_hole_punching(&mut self, ip: IpAddr, server: &Server, friend_pk: PublicKey) -> IoFuture<()> {
+        let port: u32 = 1024;
+
+        let last_punching_index = self.last_punching_index;
+        let ping_sender = (0..MAX_PORTS_TO_PUNCH)
+            .map(|i| {
+                let it = i + last_punching_index;
+                let port = port + it;
+                server.send_ping_req(
+                    &PackedNode::new(false, SocketAddr::new(ip, port as u16), &friend_pk)
+                )
+            });
+
+        let ping_stream = stream::futures_unordered(ping_sender).then(|_| Ok(()));
+
+        Box::new(ping_stream.for_each(|()| Ok(())))
+    }
+
+    // do hole punch using port guessing algorithm designed by irungentoo
+    fn punch(&mut self, ports: Vec<u16>, ip: IpAddr, server: &Server, friend_pk: PublicKey) ->IoFuture<()> {
+        if ports.is_empty() || ports.len() > MAX_CLIENTS_PER_FRIEND as usize {
+            return Box::new(future::ok(()))
+        }
+
+        // algorithm from irungentoo
+        let first_port = ports[0];
+        let num_ports = ports.len();
+        let num_same_port = ports.iter().filter(|port| **port == first_port).count();
+
+        let first_hole_punching = if num_same_port == num_ports {
+            server.send_ping_req(
+                &PackedNode::new(false, SocketAddr::new(ip, first_port), &friend_pk)
+            )
+        } else {
+            self.first_hole_punching(ports, ip, server, friend_pk)
+        };
+
+        self.first_punching_index += MAX_PORTS_TO_PUNCH;
```

Comment:
This should be inside `else` block, or even better inside `first_hole_punching` function.

Reply:
done